{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "51d32314",
   "metadata": {},
   "source": [
    "# PREPROCESSING FOR NN\n",
    "Une fois qu'on a retirer le data via le package TCGAbiolinks et utilise le SummarizedExperiment et on charge la data dans python \n",
    "\n",
    "Importation et définition du dossier de sortie: OUTDIR\n",
    "\n",
    "Checker qu’on a tout ce qu’on avait dans GDCprepare"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "51ee1f15",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "from sklearn.model_selection import GroupShuffleSplit\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.decomposition import PCA\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "OUTDIR = \"./tcga_brca_qc_report\"\n",
    "os.makedirs(OUTDIR, exist_ok=True)\n",
    "\n",
    "RANDOM_STATE = 42"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "c4ba7e2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "counts = pd.read_csv(\"brca_counts.csv\", index_col=0) #(gene x echantillon)\n",
    "meta   = pd.read_csv(\"brca_metadata.csv\", index_col=0) #(echantillon x variable)\n",
    "genes  = pd.read_csv(\"brca_genes.csv\", index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "c548434c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(60660, 1224)\n",
      "(1224, 90)\n"
     ]
    }
   ],
   "source": [
    "print(counts.shape)\n",
    "print(meta.shape)\n",
    "assert counts.shape[1] == meta.shape[0]\n",
    "\n",
    "# garder uniquement les primary tumors\n",
    "mask = (meta[\"sample_type\"] == \"Primary Tumor\").to_numpy()\n",
    "\n",
    "counts = counts.loc[:, mask]\n",
    "meta   = meta.loc[mask].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9d93831",
   "metadata": {},
   "source": [
    "ON A LES MEMES RESULTATS QUE SUR R donc c'est bon !\n",
    "Maintenant on check la metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "id": "3df18911",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "there is sample\n"
     ]
    }
   ],
   "source": [
    "#To check all the data\n",
    "meta.columns.sort_values()\n",
    "\n",
    "#To check if a column exists in the dataframe\n",
    "if 'sample' in meta.columns:\n",
    "    print(\"there is sample\")\n",
    "else:\n",
    "    print(\"no sample found\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19dc261f",
   "metadata": {},
   "source": [
    "# VERIFICATION INITIALES\n",
    "## VERIFIER LES OBJETS EN MEMOIRE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "id": "1843a7b6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'> (60660, 1111)\n",
      "<class 'pandas.core.frame.DataFrame'> (60660, 10)\n",
      "<class 'pandas.core.frame.DataFrame'> (1111, 90)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCGA-BH-A18H-01A-11R-A12D-07</th>\n",
       "      <th>TCGA-E2-A14P-01A-31R-A12D-07</th>\n",
       "      <th>TCGA-AN-A04A-01A-21R-A034-07</th>\n",
       "      <th>TCGA-5L-AAT1-01A-12R-A41B-07</th>\n",
       "      <th>TCGA-AQ-A0Y5-01A-11R-A14M-07</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003.15</th>\n",
       "      <td>426</td>\n",
       "      <td>1033</td>\n",
       "      <td>1417</td>\n",
       "      <td>929</td>\n",
       "      <td>879</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000005.6</th>\n",
       "      <td>3</td>\n",
       "      <td>21</td>\n",
       "      <td>9</td>\n",
       "      <td>3</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419.13</th>\n",
       "      <td>1439</td>\n",
       "      <td>4553</td>\n",
       "      <td>1982</td>\n",
       "      <td>933</td>\n",
       "      <td>2623</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457.14</th>\n",
       "      <td>1049</td>\n",
       "      <td>932</td>\n",
       "      <td>1227</td>\n",
       "      <td>802</td>\n",
       "      <td>1727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460.17</th>\n",
       "      <td>326</td>\n",
       "      <td>463</td>\n",
       "      <td>284</td>\n",
       "      <td>143</td>\n",
       "      <td>421</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    TCGA-BH-A18H-01A-11R-A12D-07  \\\n",
       "ENSG00000000003.15                           426   \n",
       "ENSG00000000005.6                              3   \n",
       "ENSG00000000419.13                          1439   \n",
       "ENSG00000000457.14                          1049   \n",
       "ENSG00000000460.17                           326   \n",
       "\n",
       "                    TCGA-E2-A14P-01A-31R-A12D-07  \\\n",
       "ENSG00000000003.15                          1033   \n",
       "ENSG00000000005.6                             21   \n",
       "ENSG00000000419.13                          4553   \n",
       "ENSG00000000457.14                           932   \n",
       "ENSG00000000460.17                           463   \n",
       "\n",
       "                    TCGA-AN-A04A-01A-21R-A034-07  \\\n",
       "ENSG00000000003.15                          1417   \n",
       "ENSG00000000005.6                              9   \n",
       "ENSG00000000419.13                          1982   \n",
       "ENSG00000000457.14                          1227   \n",
       "ENSG00000000460.17                           284   \n",
       "\n",
       "                    TCGA-5L-AAT1-01A-12R-A41B-07  TCGA-AQ-A0Y5-01A-11R-A14M-07  \n",
       "ENSG00000000003.15                           929                           879  \n",
       "ENSG00000000005.6                              3                             9  \n",
       "ENSG00000000419.13                           933                          2623  \n",
       "ENSG00000000457.14                           802                          1727  \n",
       "ENSG00000000460.17                           143                           421  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>patient</th>\n",
       "      <th>sample</th>\n",
       "      <th>shortLetterCode</th>\n",
       "      <th>definition</th>\n",
       "      <th>sample_submitter_id</th>\n",
       "      <th>tumor_descriptor</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>pathology_report_uuid</th>\n",
       "      <th>submitter_id</th>\n",
       "      <th>...</th>\n",
       "      <th>paper_tobacco_smoking_history</th>\n",
       "      <th>paper_CNV.Clusters</th>\n",
       "      <th>paper_Mutation.Clusters</th>\n",
       "      <th>paper_DNA.Methylation.Clusters</th>\n",
       "      <th>paper_mRNA.Clusters</th>\n",
       "      <th>paper_miRNA.Clusters</th>\n",
       "      <th>paper_lncRNA.Clusters</th>\n",
       "      <th>paper_Protein.Clusters</th>\n",
       "      <th>paper_PARADIGM.Clusters</th>\n",
       "      <th>paper_Pan.Gyn.Clusters</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-BH-A18H-01A-11R-A12D-07</td>\n",
       "      <td>TCGA-BH-A18H</td>\n",
       "      <td>TCGA-BH-A18H-01A</td>\n",
       "      <td>TP</td>\n",
       "      <td>Primary solid Tumor</td>\n",
       "      <td>TCGA-BH-A18H-01A</td>\n",
       "      <td>Primary</td>\n",
       "      <td>9321161b-d568-4d38-aed3-956786a06329</td>\n",
       "      <td>DF536409-601F-489B-9F8A-A2B372BF9999</td>\n",
       "      <td>TCGA-BH-A18H</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C4</td>\n",
       "      <td>C4</td>\n",
       "      <td>C1</td>\n",
       "      <td>C2</td>\n",
       "      <td>C3</td>\n",
       "      <td>C1</td>\n",
       "      <td>C1</td>\n",
       "      <td>C6</td>\n",
       "      <td>C3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-E2-A14P-01A-31R-A12D-07</td>\n",
       "      <td>TCGA-E2-A14P</td>\n",
       "      <td>TCGA-E2-A14P-01A</td>\n",
       "      <td>TP</td>\n",
       "      <td>Primary solid Tumor</td>\n",
       "      <td>TCGA-E2-A14P-01A</td>\n",
       "      <td>Primary</td>\n",
       "      <td>58bfe278-a80f-4722-b286-4d966214d244</td>\n",
       "      <td>FD8CDFD9-6DCF-4CE1-9BB1-A7FF568EE53C</td>\n",
       "      <td>TCGA-E2-A14P</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C4</td>\n",
       "      <td>C4</td>\n",
       "      <td>C1</td>\n",
       "      <td>C2</td>\n",
       "      <td>C7</td>\n",
       "      <td>C2</td>\n",
       "      <td>C1</td>\n",
       "      <td>C4</td>\n",
       "      <td>C3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-AN-A04A-01A-21R-A034-07</td>\n",
       "      <td>TCGA-AN-A04A</td>\n",
       "      <td>TCGA-AN-A04A-01A</td>\n",
       "      <td>TP</td>\n",
       "      <td>Primary solid Tumor</td>\n",
       "      <td>TCGA-AN-A04A-01A</td>\n",
       "      <td>Primary</td>\n",
       "      <td>1d8f72b7-8596-40e8-afdf-244dd897758c</td>\n",
       "      <td>6FB7A794-C3E7-4501-A844-CF65B9C3D32C</td>\n",
       "      <td>TCGA-AN-A04A</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C6</td>\n",
       "      <td>C1</td>\n",
       "      <td>C1</td>\n",
       "      <td>C1</td>\n",
       "      <td>C3</td>\n",
       "      <td>C2</td>\n",
       "      <td>C1</td>\n",
       "      <td>C5</td>\n",
       "      <td>C5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-5L-AAT1-01A-12R-A41B-07</td>\n",
       "      <td>TCGA-5L-AAT1</td>\n",
       "      <td>TCGA-5L-AAT1-01A</td>\n",
       "      <td>TP</td>\n",
       "      <td>Primary solid Tumor</td>\n",
       "      <td>TCGA-5L-AAT1-01A</td>\n",
       "      <td>Primary</td>\n",
       "      <td>a31dec9b-3626-4cb0-8265-1425d5ce1c1e</td>\n",
       "      <td>B5CA42BB-9514-42C6-9FB0-C8889C1DC51A</td>\n",
       "      <td>TCGA-5L-AAT1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C9</td>\n",
       "      <td>C1</td>\n",
       "      <td>C2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-AQ-A0Y5-01A-11R-A14M-07</td>\n",
       "      <td>TCGA-AQ-A0Y5</td>\n",
       "      <td>TCGA-AQ-A0Y5-01A</td>\n",
       "      <td>TP</td>\n",
       "      <td>Primary solid Tumor</td>\n",
       "      <td>TCGA-AQ-A0Y5-01A</td>\n",
       "      <td>Primary</td>\n",
       "      <td>f5675b71-ed1b-4fdb-be07-d627dc98ed88</td>\n",
       "      <td>5BC01981-366E-44EC-A986-087DB7904E5C</td>\n",
       "      <td>TCGA-AQ-A0Y5</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C4</td>\n",
       "      <td>C4</td>\n",
       "      <td>C2</td>\n",
       "      <td>C1</td>\n",
       "      <td>C2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C5</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 90 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        barcode       patient            sample  \\\n",
       "0  TCGA-BH-A18H-01A-11R-A12D-07  TCGA-BH-A18H  TCGA-BH-A18H-01A   \n",
       "1  TCGA-E2-A14P-01A-31R-A12D-07  TCGA-E2-A14P  TCGA-E2-A14P-01A   \n",
       "2  TCGA-AN-A04A-01A-21R-A034-07  TCGA-AN-A04A  TCGA-AN-A04A-01A   \n",
       "3  TCGA-5L-AAT1-01A-12R-A41B-07  TCGA-5L-AAT1  TCGA-5L-AAT1-01A   \n",
       "4  TCGA-AQ-A0Y5-01A-11R-A14M-07  TCGA-AQ-A0Y5  TCGA-AQ-A0Y5-01A   \n",
       "\n",
       "  shortLetterCode           definition sample_submitter_id tumor_descriptor  \\\n",
       "0              TP  Primary solid Tumor    TCGA-BH-A18H-01A          Primary   \n",
       "1              TP  Primary solid Tumor    TCGA-E2-A14P-01A          Primary   \n",
       "2              TP  Primary solid Tumor    TCGA-AN-A04A-01A          Primary   \n",
       "3              TP  Primary solid Tumor    TCGA-5L-AAT1-01A          Primary   \n",
       "4              TP  Primary solid Tumor    TCGA-AQ-A0Y5-01A          Primary   \n",
       "\n",
       "                              sample_id                 pathology_report_uuid  \\\n",
       "0  9321161b-d568-4d38-aed3-956786a06329  DF536409-601F-489B-9F8A-A2B372BF9999   \n",
       "1  58bfe278-a80f-4722-b286-4d966214d244  FD8CDFD9-6DCF-4CE1-9BB1-A7FF568EE53C   \n",
       "2  1d8f72b7-8596-40e8-afdf-244dd897758c  6FB7A794-C3E7-4501-A844-CF65B9C3D32C   \n",
       "3  a31dec9b-3626-4cb0-8265-1425d5ce1c1e  B5CA42BB-9514-42C6-9FB0-C8889C1DC51A   \n",
       "4  f5675b71-ed1b-4fdb-be07-d627dc98ed88  5BC01981-366E-44EC-A986-087DB7904E5C   \n",
       "\n",
       "   submitter_id  ... paper_tobacco_smoking_history paper_CNV.Clusters  \\\n",
       "0  TCGA-BH-A18H  ...                           NaN                 C4   \n",
       "1  TCGA-E2-A14P  ...                           NaN                 C4   \n",
       "2  TCGA-AN-A04A  ...                           NaN                 C6   \n",
       "3  TCGA-5L-AAT1  ...                           NaN                NaN   \n",
       "4  TCGA-AQ-A0Y5  ...                           NaN                 C4   \n",
       "\n",
       "   paper_Mutation.Clusters paper_DNA.Methylation.Clusters  \\\n",
       "0                       C4                             C1   \n",
       "1                       C4                             C1   \n",
       "2                       C1                             C1   \n",
       "3                       C9                             C1   \n",
       "4                       C4                             C2   \n",
       "\n",
       "   paper_mRNA.Clusters paper_miRNA.Clusters paper_lncRNA.Clusters  \\\n",
       "0                   C2                   C3                    C1   \n",
       "1                   C2                   C7                    C2   \n",
       "2                   C1                   C3                    C2   \n",
       "3                   C2                  NaN                   NaN   \n",
       "4                   C1                   C2                   NaN   \n",
       "\n",
       "  paper_Protein.Clusters paper_PARADIGM.Clusters  paper_Pan.Gyn.Clusters  \n",
       "0                     C1                      C6                      C3  \n",
       "1                     C1                      C4                      C3  \n",
       "2                     C1                      C5                      C5  \n",
       "3                     C2                     NaN                      C1  \n",
       "4                    NaN                      C5                     NaN  \n",
       "\n",
       "[5 rows x 90 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>score</th>\n",
       "      <th>phase</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>level</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>havana_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003.15</th>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000000003.15</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:11858</td>\n",
       "      <td>OTTHUMG00000022002.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000005.6</th>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000000005.6</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TNMD</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:17757</td>\n",
       "      <td>OTTHUMG00000022001.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419.13</th>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000000419.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:3005</td>\n",
       "      <td>OTTHUMG00000032742.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457.14</th>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000000457.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SCYL3</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:19285</td>\n",
       "      <td>OTTHUMG00000035941.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460.17</th>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:25565</td>\n",
       "      <td>OTTHUMG00000035821.9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    source  type  score  phase             gene_id  \\\n",
       "ENSG00000000003.15  HAVANA  gene    NaN    NaN  ENSG00000000003.15   \n",
       "ENSG00000000005.6   HAVANA  gene    NaN    NaN   ENSG00000000005.6   \n",
       "ENSG00000000419.13  HAVANA  gene    NaN    NaN  ENSG00000000419.13   \n",
       "ENSG00000000457.14  HAVANA  gene    NaN    NaN  ENSG00000000457.14   \n",
       "ENSG00000000460.17  HAVANA  gene    NaN    NaN  ENSG00000000460.17   \n",
       "\n",
       "                         gene_type gene_name  level     hgnc_id  \\\n",
       "ENSG00000000003.15  protein_coding    TSPAN6      2  HGNC:11858   \n",
       "ENSG00000000005.6   protein_coding      TNMD      2  HGNC:17757   \n",
       "ENSG00000000419.13  protein_coding      DPM1      2   HGNC:3005   \n",
       "ENSG00000000457.14  protein_coding     SCYL3      2  HGNC:19285   \n",
       "ENSG00000000460.17  protein_coding  C1orf112      2  HGNC:25565   \n",
       "\n",
       "                             havana_gene  \n",
       "ENSG00000000003.15  OTTHUMG00000022002.2  \n",
       "ENSG00000000005.6   OTTHUMG00000022001.2  \n",
       "ENSG00000000419.13  OTTHUMG00000032742.2  \n",
       "ENSG00000000457.14  OTTHUMG00000035941.6  \n",
       "ENSG00000000460.17  OTTHUMG00000035821.9  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Attendu: counts, genes, meta existent déjà (issus de GDCprepare)\n",
    "print(type(counts), counts.shape)\n",
    "print(type(genes), genes.shape)\n",
    "print(type(meta), meta.shape)\n",
    "\n",
    "display(counts.iloc[:5, :5])\n",
    "display(meta.head())\n",
    "display(genes.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2fff2c1",
   "metadata": {},
   "source": [
    "## DEFINIR LES COLONNES CLES \n",
    "On défini sample_col et patient_col car TCGA a plusieurs niveaux d’identifiants et on veut être sûr qu’on a le bon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "id": "5d17a82a",
   "metadata": {},
   "outputs": [],
   "source": [
    "SAMPLE_COL  = \"barcode\"   # colonne meta qui correspond aux colonnes de counts\n",
    "PATIENT_COL = \"patient\"  # colonne meta pour grouper les échantillons par patient\n",
    "\n",
    "if SAMPLE_COL not in meta.columns:\n",
    "    raise ValueError(f\"{SAMPLE_COL} absent de meta.columns. Colonnes disponibles: {list(meta.columns)[:30]} ...\")\n",
    "\n",
    "if PATIENT_COL not in meta.columns:\n",
    "    raise ValueError(f\"{PATIENT_COL} absent de meta.columns. Colonnes disponibles: {list(meta.columns)[:30]} ...\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f40e96e5",
   "metadata": {},
   "source": [
    "## ALIGNER LES COUNTS ET LES META \n",
    "Réordonne/Aligne: intersection entre counts.columns et meta[SAMPLE_COL]\n",
    "\n",
    "Pour que chaque colonne de counts correspondent à une ligne de meta "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "id": "471789f9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Aligned counts: (60660, 1111)\n",
      "Aligned meta: (1111, 90)\n"
     ]
    }
   ],
   "source": [
    "# counts: genes x samples\n",
    "if counts.index.duplicated().any():\n",
    "    raise ValueError(\"counts.index contient des doublons (gènes). À corriger avant QC.\")\n",
    "\n",
    "if counts.columns.duplicated().any():\n",
    "    raise ValueError(\"counts.columns contient des doublons (échantillons). À corriger avant QC.\")\n",
    "\n",
    "# Intersection\n",
    "meta_samples = set(meta[SAMPLE_COL].astype(str))\n",
    "common = [s for s in counts.columns.astype(str) if s in meta_samples]\n",
    "\n",
    "if len(common) < 10:\n",
    "    raise ValueError(\"Trop peu d'échantillons en commun entre counts et meta. Vérifie les IDs.\")\n",
    "\n",
    "counts = counts.loc[:, common].copy()\n",
    "\n",
    "# Meta réordonnée selon counts\n",
    "meta = meta.copy()\n",
    "meta[SAMPLE_COL] = meta[SAMPLE_COL].astype(str)\n",
    "meta = meta[meta[SAMPLE_COL].isin(common)].set_index(SAMPLE_COL).loc[common].reset_index()\n",
    "\n",
    "print(\"Aligned counts:\", counts.shape)\n",
    "print(\"Aligned meta:\", meta.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f58d3745",
   "metadata": {},
   "source": [
    "## VERIFIER LA COHERENCE AVEC LES GENES\n",
    "On vérifie si genes.index == counts.index → chaque ligne de counts correspond bien à un gène défini"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "id": "d462bea7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OK: genes.index == counts.index\n"
     ]
    }
   ],
   "source": [
    "# On suppose que genes a une colonne qui correspond à counts.index (ex: gene_id / ensembl_gene_id)\n",
    "# Si gene est déjà indexé par le bon ID, laisser comme ça.\n",
    "if genes.index.equals(counts.index):\n",
    "    print(\"OK: genes.index == counts.index\")\n",
    "else:\n",
    "    # sinon, mapper une colonne\n",
    "    # -> adapter GENE_ID_COL si nécessaire -> pas besoin ici gene.index == counts.index\n",
    "    GENE_ID_COL = None  # ex: \"gene_id\" ou \"ensembl_gene_id\"\n",
    "    if GENE_ID_COL is None:\n",
    "        print(\"NOTE: genes.index ne matche pas counts.index. Si tu veux joindre l'annotation, définis GENE_ID_COL.\")\n",
    "    else:\n",
    "        if GENE_ID_COL not in genes.columns:\n",
    "            raise ValueError(f\"{GENE_ID_COL} absent de genes.columns.\")\n",
    "        genes = genes.set_index(GENE_ID_COL)\n",
    "        genes = genes.loc[counts.index]\n",
    "        print(\"genes réaligné sur counts.index:\", genes.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2abb4a47",
   "metadata": {},
   "source": [
    "# SPLIT PAR GROUPE PATIENT\n",
    "On split 80/20 avc GroupeShuffleSplit, on groupe par patient_id et on vérifie qu’aucun patient n’est dans les deux sets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "id": "0f07fd10",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Train: (60660, 889) Test: (60660, 222)\n"
     ]
    }
   ],
   "source": [
    "gss = GroupShuffleSplit(n_splits=1, test_size=0.2, random_state=RANDOM_STATE)\n",
    "\n",
    "idx = np.arange(meta.shape[0])\n",
    "groups = meta[PATIENT_COL].astype(str).values\n",
    "\n",
    "train_idx, test_idx = next(gss.split(idx, groups=groups))\n",
    "\n",
    "train_samples = meta.loc[train_idx, SAMPLE_COL].tolist()\n",
    "test_samples  = meta.loc[test_idx,  SAMPLE_COL].tolist()\n",
    "\n",
    "# Sanity check: pas de patient en commun\n",
    "train_pat = set(meta.loc[train_idx, PATIENT_COL].astype(str))\n",
    "test_pat  = set(meta.loc[test_idx,  PATIENT_COL].astype(str))\n",
    "overlap = train_pat & test_pat\n",
    "if overlap:\n",
    "    raise RuntimeError(f\"Leakage patient détecté: {len(overlap)} patients se retrouvent dans train et test.\")\n",
    "\n",
    "counts_tr = counts[train_samples]\n",
    "counts_te = counts[test_samples]\n",
    "\n",
    "meta_tr = meta.set_index(SAMPLE_COL).loc[train_samples].reset_index()\n",
    "meta_te = meta.set_index(SAMPLE_COL).loc[test_samples].reset_index()\n",
    "\n",
    "print(\"Train:\", counts_tr.shape, \"Test:\", counts_te.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eed0a2d3",
   "metadata": {},
   "source": [
    "# QC Calcul des métrique QC\n",
    "On calcul, pour chaque sample:\n",
    "\n",
    "- *libsize*: somme des counts\n",
    "- *detected_genes*: nbr de gène exprimés\n",
    "- *zero_fraction:* proportion de zéros\n",
    "- *top1_frac*: dominance du gène le plus exprimé, top1 est une coordonnée PCA par échantillon.\n",
    "\n",
    "Pour qu’avec ces Qc on détecte les échantillons de mauvaise qualité, problème technique et les outliers évidents"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "id": "9bf4bced",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>libsize</th>\n",
       "      <th>detected_genes</th>\n",
       "      <th>zero_fraction</th>\n",
       "      <th>top1_frac</th>\n",
       "      <th>patient</th>\n",
       "      <th>sample</th>\n",
       "      <th>shortLetterCode</th>\n",
       "      <th>definition</th>\n",
       "      <th>sample_submitter_id</th>\n",
       "      <th>tumor_descriptor</th>\n",
       "      <th>...</th>\n",
       "      <th>paper_tobacco_smoking_history</th>\n",
       "      <th>paper_CNV.Clusters</th>\n",
       "      <th>paper_Mutation.Clusters</th>\n",
       "      <th>paper_DNA.Methylation.Clusters</th>\n",
       "      <th>paper_mRNA.Clusters</th>\n",
       "      <th>paper_miRNA.Clusters</th>\n",
       "      <th>paper_lncRNA.Clusters</th>\n",
       "      <th>paper_Protein.Clusters</th>\n",
       "      <th>paper_PARADIGM.Clusters</th>\n",
       "      <th>paper_Pan.Gyn.Clusters</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-BH-A18H-01A-11R-A12D-07</th>\n",
       "      <td>41748828.0</td>\n",
       "      <td>31328.0</td>\n",
       "      <td>0.483548</td>\n",
       "      <td>0.016142</td>\n",
       "      <td>TCGA-BH-A18H</td>\n",
       "      <td>TCGA-BH-A18H-01A</td>\n",
       "      <td>TP</td>\n",
       "      <td>Primary solid Tumor</td>\n",
       "      <td>TCGA-BH-A18H-01A</td>\n",
       "      <td>Primary</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C4</td>\n",
       "      <td>C4</td>\n",
       "      <td>C1</td>\n",
       "      <td>C2</td>\n",
       "      <td>C3</td>\n",
       "      <td>C1</td>\n",
       "      <td>C1</td>\n",
       "      <td>C6</td>\n",
       "      <td>C3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-AN-A04A-01A-21R-A034-07</th>\n",
       "      <td>68281533.0</td>\n",
       "      <td>30936.0</td>\n",
       "      <td>0.490010</td>\n",
       "      <td>0.017330</td>\n",
       "      <td>TCGA-AN-A04A</td>\n",
       "      <td>TCGA-AN-A04A-01A</td>\n",
       "      <td>TP</td>\n",
       "      <td>Primary solid Tumor</td>\n",
       "      <td>TCGA-AN-A04A-01A</td>\n",
       "      <td>Primary</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C6</td>\n",
       "      <td>C1</td>\n",
       "      <td>C1</td>\n",
       "      <td>C1</td>\n",
       "      <td>C3</td>\n",
       "      <td>C2</td>\n",
       "      <td>C1</td>\n",
       "      <td>C5</td>\n",
       "      <td>C5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-5L-AAT1-01A-12R-A41B-07</th>\n",
       "      <td>24260663.0</td>\n",
       "      <td>29142.0</td>\n",
       "      <td>0.519585</td>\n",
       "      <td>0.014799</td>\n",
       "      <td>TCGA-5L-AAT1</td>\n",
       "      <td>TCGA-5L-AAT1-01A</td>\n",
       "      <td>TP</td>\n",
       "      <td>Primary solid Tumor</td>\n",
       "      <td>TCGA-5L-AAT1-01A</td>\n",
       "      <td>Primary</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C9</td>\n",
       "      <td>C1</td>\n",
       "      <td>C2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-AQ-A0Y5-01A-11R-A14M-07</th>\n",
       "      <td>63988272.0</td>\n",
       "      <td>32629.0</td>\n",
       "      <td>0.462100</td>\n",
       "      <td>0.012445</td>\n",
       "      <td>TCGA-AQ-A0Y5</td>\n",
       "      <td>TCGA-AQ-A0Y5-01A</td>\n",
       "      <td>TP</td>\n",
       "      <td>Primary solid Tumor</td>\n",
       "      <td>TCGA-AQ-A0Y5-01A</td>\n",
       "      <td>Primary</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C4</td>\n",
       "      <td>C4</td>\n",
       "      <td>C2</td>\n",
       "      <td>C1</td>\n",
       "      <td>C2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C5</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-E9-A1R4-01A-21R-A14D-07</th>\n",
       "      <td>50077273.0</td>\n",
       "      <td>33222.0</td>\n",
       "      <td>0.452324</td>\n",
       "      <td>0.013934</td>\n",
       "      <td>TCGA-E9-A1R4</td>\n",
       "      <td>TCGA-E9-A1R4-01A</td>\n",
       "      <td>TP</td>\n",
       "      <td>Primary solid Tumor</td>\n",
       "      <td>TCGA-E9-A1R4-01A</td>\n",
       "      <td>Primary</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C6</td>\n",
       "      <td>C9</td>\n",
       "      <td>C1</td>\n",
       "      <td>C1</td>\n",
       "      <td>C3</td>\n",
       "      <td>C1</td>\n",
       "      <td>C1</td>\n",
       "      <td>C6</td>\n",
       "      <td>C3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 93 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 libsize  detected_genes  zero_fraction  \\\n",
       "TCGA-BH-A18H-01A-11R-A12D-07  41748828.0         31328.0       0.483548   \n",
       "TCGA-AN-A04A-01A-21R-A034-07  68281533.0         30936.0       0.490010   \n",
       "TCGA-5L-AAT1-01A-12R-A41B-07  24260663.0         29142.0       0.519585   \n",
       "TCGA-AQ-A0Y5-01A-11R-A14M-07  63988272.0         32629.0       0.462100   \n",
       "TCGA-E9-A1R4-01A-21R-A14D-07  50077273.0         33222.0       0.452324   \n",
       "\n",
       "                              top1_frac       patient            sample  \\\n",
       "TCGA-BH-A18H-01A-11R-A12D-07   0.016142  TCGA-BH-A18H  TCGA-BH-A18H-01A   \n",
       "TCGA-AN-A04A-01A-21R-A034-07   0.017330  TCGA-AN-A04A  TCGA-AN-A04A-01A   \n",
       "TCGA-5L-AAT1-01A-12R-A41B-07   0.014799  TCGA-5L-AAT1  TCGA-5L-AAT1-01A   \n",
       "TCGA-AQ-A0Y5-01A-11R-A14M-07   0.012445  TCGA-AQ-A0Y5  TCGA-AQ-A0Y5-01A   \n",
       "TCGA-E9-A1R4-01A-21R-A14D-07   0.013934  TCGA-E9-A1R4  TCGA-E9-A1R4-01A   \n",
       "\n",
       "                             shortLetterCode           definition  \\\n",
       "TCGA-BH-A18H-01A-11R-A12D-07              TP  Primary solid Tumor   \n",
       "TCGA-AN-A04A-01A-21R-A034-07              TP  Primary solid Tumor   \n",
       "TCGA-5L-AAT1-01A-12R-A41B-07              TP  Primary solid Tumor   \n",
       "TCGA-AQ-A0Y5-01A-11R-A14M-07              TP  Primary solid Tumor   \n",
       "TCGA-E9-A1R4-01A-21R-A14D-07              TP  Primary solid Tumor   \n",
       "\n",
       "                             sample_submitter_id tumor_descriptor  ...  \\\n",
       "TCGA-BH-A18H-01A-11R-A12D-07    TCGA-BH-A18H-01A          Primary  ...   \n",
       "TCGA-AN-A04A-01A-21R-A034-07    TCGA-AN-A04A-01A          Primary  ...   \n",
       "TCGA-5L-AAT1-01A-12R-A41B-07    TCGA-5L-AAT1-01A          Primary  ...   \n",
       "TCGA-AQ-A0Y5-01A-11R-A14M-07    TCGA-AQ-A0Y5-01A          Primary  ...   \n",
       "TCGA-E9-A1R4-01A-21R-A14D-07    TCGA-E9-A1R4-01A          Primary  ...   \n",
       "\n",
       "                             paper_tobacco_smoking_history paper_CNV.Clusters  \\\n",
       "TCGA-BH-A18H-01A-11R-A12D-07                           NaN                 C4   \n",
       "TCGA-AN-A04A-01A-21R-A034-07                           NaN                 C6   \n",
       "TCGA-5L-AAT1-01A-12R-A41B-07                           NaN                NaN   \n",
       "TCGA-AQ-A0Y5-01A-11R-A14M-07                           NaN                 C4   \n",
       "TCGA-E9-A1R4-01A-21R-A14D-07                           NaN                 C6   \n",
       "\n",
       "                             paper_Mutation.Clusters  \\\n",
       "TCGA-BH-A18H-01A-11R-A12D-07                      C4   \n",
       "TCGA-AN-A04A-01A-21R-A034-07                      C1   \n",
       "TCGA-5L-AAT1-01A-12R-A41B-07                      C9   \n",
       "TCGA-AQ-A0Y5-01A-11R-A14M-07                      C4   \n",
       "TCGA-E9-A1R4-01A-21R-A14D-07                      C9   \n",
       "\n",
       "                             paper_DNA.Methylation.Clusters  \\\n",
       "TCGA-BH-A18H-01A-11R-A12D-07                             C1   \n",
       "TCGA-AN-A04A-01A-21R-A034-07                             C1   \n",
       "TCGA-5L-AAT1-01A-12R-A41B-07                             C1   \n",
       "TCGA-AQ-A0Y5-01A-11R-A14M-07                             C2   \n",
       "TCGA-E9-A1R4-01A-21R-A14D-07                             C1   \n",
       "\n",
       "                             paper_mRNA.Clusters  paper_miRNA.Clusters  \\\n",
       "TCGA-BH-A18H-01A-11R-A12D-07                  C2                    C3   \n",
       "TCGA-AN-A04A-01A-21R-A034-07                  C1                    C3   \n",
       "TCGA-5L-AAT1-01A-12R-A41B-07                  C2                   NaN   \n",
       "TCGA-AQ-A0Y5-01A-11R-A14M-07                  C1                    C2   \n",
       "TCGA-E9-A1R4-01A-21R-A14D-07                  C1                    C3   \n",
       "\n",
       "                             paper_lncRNA.Clusters  paper_Protein.Clusters  \\\n",
       "TCGA-BH-A18H-01A-11R-A12D-07                    C1                      C1   \n",
       "TCGA-AN-A04A-01A-21R-A034-07                    C2                      C1   \n",
       "TCGA-5L-AAT1-01A-12R-A41B-07                   NaN                      C2   \n",
       "TCGA-AQ-A0Y5-01A-11R-A14M-07                   NaN                     NaN   \n",
       "TCGA-E9-A1R4-01A-21R-A14D-07                    C1                      C1   \n",
       "\n",
       "                             paper_PARADIGM.Clusters paper_Pan.Gyn.Clusters  \n",
       "TCGA-BH-A18H-01A-11R-A12D-07                      C6                     C3  \n",
       "TCGA-AN-A04A-01A-21R-A034-07                      C5                     C5  \n",
       "TCGA-5L-AAT1-01A-12R-A41B-07                     NaN                     C1  \n",
       "TCGA-AQ-A0Y5-01A-11R-A14M-07                      C5                    NaN  \n",
       "TCGA-E9-A1R4-01A-21R-A14D-07                      C6                     C3  \n",
       "\n",
       "[5 rows x 93 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def sample_qc_metrics(counts_gxs: pd.DataFrame) -> pd.DataFrame:\n",
    "    libsize = counts_gxs.sum(axis=0).astype(float)\n",
    "    detected = (counts_gxs > 0).sum(axis=0).astype(float)\n",
    "    zero_frac = (counts_gxs == 0).mean(axis=0).astype(float)\n",
    "    top1 = (counts_gxs.max(axis=0) / libsize.replace(0, np.nan)).fillna(0.0).astype(float)\n",
    "    return pd.DataFrame({\n",
    "        \"libsize\": libsize,\n",
    "        \"detected_genes\": detected,\n",
    "        \"zero_fraction\": zero_frac,\n",
    "        \"top1_frac\": top1,\n",
    "    })\n",
    "\n",
    "qc_tr = sample_qc_metrics(counts_tr)\n",
    "qc_te = sample_qc_metrics(counts_te)\n",
    "\n",
    "qc_tr = qc_tr.join(meta_tr.set_index(SAMPLE_COL), how=\"left\")\n",
    "qc_te = qc_te.join(meta_te.set_index(SAMPLE_COL), how=\"left\")\n",
    "\n",
    "display(qc_tr.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19e328cf",
   "metadata": {},
   "source": [
    "# QC- Seuil appris sur Train\n",
    "## Fonction seuil robuste\n",
    "On définit median ± k*MAD (*Median Absolute Deviation)*\n",
    "\n",
    "Parce que les distribution RNA-seq ne sont pas normales, MAD est robustes aux outliers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "bac85e01",
   "metadata": {},
   "outputs": [],
   "source": [
    "def mad_bounds(x: pd.Series, k: float = 3.5):\n",
    "    x = pd.to_numeric(x, errors=\"coerce\").dropna().astype(float)\n",
    "    med = np.median(x)\n",
    "    mad = np.median(np.abs(x - med))\n",
    "    robust_sd = 1.4826 * mad if mad > 0 else np.std(x)\n",
    "    lo = med - k * robust_sd\n",
    "    hi = med + k * robust_sd\n",
    "    return lo, hi"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7dd4496a",
   "metadata": {},
   "source": [
    "## Apprentissage/définition des seuils QC sur train ( à faire sur test après)\n",
    "On calcul les seuils QC uniqmt sur train, on ajoute un plancher (garde fou biologique)\n",
    "\n",
    "Cela évite toute information pour test, les seuils deviennent des hyper-paramètre appris\n",
    "\n",
    "Bulk RNA-seq humain :\n",
    "~15 000–20 000 gènes détectés"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "id": "0a5547ae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'libsize_lo': 3643916.172600001,\n",
       " 'libsize_hi': 111782829.8274,\n",
       " 'detected_lo': 25803.17,\n",
       " 'detected_hi': 39294.83,\n",
       " 'zero_lo': 0.3522118364655456,\n",
       " 'zero_hi': 0.5746262776129245,\n",
       " 'top1_lo': -0.009222292377061271,\n",
       " 'top1_hi': 0.03635283312095532}"
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lib_lo, lib_hi = mad_bounds(qc_tr[\"libsize\"], k=3.5)\n",
    "det_lo, det_hi = mad_bounds(qc_tr[\"detected_genes\"], k=3.5) #nbr de gènes détectés par échantillon\n",
    "zf_lo,  zf_hi  = mad_bounds(qc_tr[\"zero_fraction\"], k=3.5)\n",
    "t1_lo,  t1_hi  = mad_bounds(qc_tr[\"top1_frac\"], k=3.5) #top1 = projection de chaque échantillon sur la 1ʳᵉ composante principale (PC1).\n",
    "\n",
    "# Option: planchers/plafonds de sécurité (à ajuster selon dataset)\n",
    "lib_lo = max(lib_lo, 1e6)     # On veut au moins 1 million de lectures/read par échantillon (en dessous on a du bruit technique dominant)\n",
    "det_lo = max(det_lo, 10000)    # On elimine les échantillons avec moins de 10000 gènes détectés, car grand dataset (+1000 échantillons)\n",
    "zf_hi  = min(zf_hi, 0.95)       # on autorise au maximum 5 % des échantillons les plus extrêmes, quelques outliers abbérants peuvent gonfler artificiellement le seuils oui annuler l'effet de QC.\n",
    "\n",
    "qc_thresholds = {\n",
    "    \"libsize_lo\": lib_lo, \"libsize_hi\": lib_hi,\n",
    "    \"detected_lo\": det_lo, \"detected_hi\": det_hi,\n",
    "    \"zero_lo\": zf_lo, \"zero_hi\": zf_hi,\n",
    "    \"top1_lo\": t1_lo, \"top1_hi\": t1_hi\n",
    "}\n",
    "\n",
    "qc_thresholds"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc163a7f",
   "metadata": {},
   "source": [
    "## Application des seuils au Test dataset\n",
    "ON applique ces mêmes seuil au test, les quelques samples exclus sont gardé dans la colonnes crée keep."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "id": "b8542363",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Train kept: 814 / 889\n",
      "Test  kept: 208 / 222\n"
     ]
    }
   ],
   "source": [
    "def apply_qc(df_qc: pd.DataFrame, thr: dict) -> pd.Series:\n",
    "    m = pd.Series(True, index=df_qc.index)\n",
    "    m &= df_qc[\"libsize\"].between(thr[\"libsize_lo\"], thr[\"libsize_hi\"])\n",
    "    m &= df_qc[\"detected_genes\"].between(thr[\"detected_lo\"], thr[\"detected_hi\"])\n",
    "    m &= df_qc[\"zero_fraction\"].between(thr[\"zero_lo\"], thr[\"zero_hi\"])\n",
    "    m &= df_qc[\"top1_frac\"].between(thr[\"top1_lo\"], thr[\"top1_hi\"])\n",
    "    return m\n",
    "\n",
    "keep_tr = apply_qc(qc_tr, qc_thresholds)\n",
    "keep_te = apply_qc(qc_te, qc_thresholds)\n",
    "\n",
    "print(\"Train kept:\", int(keep_tr.sum()), \"/\", len(keep_tr))\n",
    "print(\"Test  kept:\", int(keep_te.sum()), \"/\", len(keep_te))\n",
    "\n",
    "qc_tr_out = qc_tr.copy()\n",
    "qc_te_out = qc_te.copy()\n",
    "qc_tr_out[\"keep\"] = keep_tr.values\n",
    "qc_te_out[\"keep\"] = keep_te.values"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6e39c8b",
   "metadata": {},
   "source": [
    "# GRAPHS POUR QC\n",
    "## SCATTER + HIST + LIGENS SEUIL\n",
    "Ce sont les fonctions pour histogrames, scatterplots et sauvegarde automatique."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "id": "86db5183",
   "metadata": {},
   "outputs": [],
   "source": [
    "def savefig(name):\n",
    "    path = os.path.join(OUTDIR, name)\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(path, dpi=160, bbox_inches=\"tight\")\n",
    "    plt.close()\n",
    "    return path\n",
    "\n",
    "def qc_hist(train_df, test_df, col, thr=None, bins=60):\n",
    "    plt.figure()\n",
    "    plt.hist(train_df[col].dropna(), bins=bins, alpha=0.6, label=\"train\")\n",
    "    plt.hist(test_df[col].dropna(), bins=bins, alpha=0.6, label=\"test\")\n",
    "    plt.xlabel(col); plt.ylabel(\"n\")\n",
    "    if thr is not None:\n",
    "        lo, hi = thr\n",
    "        plt.axvline(lo)\n",
    "        plt.axvline(hi)\n",
    "    plt.legend(frameon=False)\n",
    "    return savefig(f\"qc_hist_{col}.png\")\n",
    "\n",
    "def qc_scatter(df, x, y, keep_col=\"keep\", title=\"\"):\n",
    "    plt.figure()\n",
    "    if keep_col in df.columns:\n",
    "        m = df[keep_col].astype(bool).values\n",
    "        plt.scatter(df.loc[m, x], df.loc[m, y], s=12, alpha=0.7, label=\"kept\")\n",
    "        plt.scatter(df.loc[~m, x], df.loc[~m, y], s=12, alpha=0.7, label=\"filtered\")\n",
    "        plt.legend(frameon=False)\n",
    "    else:\n",
    "        plt.scatter(df[x], df[y], s=12, alpha=0.7)\n",
    "    plt.xlabel(x); plt.ylabel(y); plt.title(title)\n",
    "    return savefig(f\"qc_scatter_{x}_vs_{y}_{title.replace(' ', '_')}.png\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49b1300c",
   "metadata": {},
   "source": [
    "## QC 1, 2, 3, 4 : LEURS DISTRIBUTION ET SEUIL\n",
    "Histograme train vs test et ajout des lignes de seuil\n",
    "\n",
    "Pour vérifier que les seuils sont raisonnables et que train & test ont des distribution comparable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "id": "54ace541",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>libsize</th>\n",
       "      <th>detected_genes</th>\n",
       "      <th>zero_fraction</th>\n",
       "      <th>top1_frac</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>8.890000e+02</td>\n",
       "      <td>889.000000</td>\n",
       "      <td>889.000000</td>\n",
       "      <td>889.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>5.771642e+07</td>\n",
       "      <td>32537.221597</td>\n",
       "      <td>0.463613</td>\n",
       "      <td>0.017842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>1.466917e+07</td>\n",
       "      <td>2529.492654</td>\n",
       "      <td>0.041700</td>\n",
       "      <td>0.013975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>1.922512e+07</td>\n",
       "      <td>24322.000000</td>\n",
       "      <td>0.098566</td>\n",
       "      <td>0.004829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>4.734575e+07</td>\n",
       "      <td>31190.000000</td>\n",
       "      <td>0.442598</td>\n",
       "      <td>0.009936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>5.771337e+07</td>\n",
       "      <td>32549.000000</td>\n",
       "      <td>0.463419</td>\n",
       "      <td>0.013565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>6.817859e+07</td>\n",
       "      <td>33812.000000</td>\n",
       "      <td>0.485823</td>\n",
       "      <td>0.019978</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>1.140157e+08</td>\n",
       "      <td>54681.000000</td>\n",
       "      <td>0.599044</td>\n",
       "      <td>0.105110</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            libsize  detected_genes  zero_fraction   top1_frac\n",
       "count  8.890000e+02      889.000000     889.000000  889.000000\n",
       "mean   5.771642e+07    32537.221597       0.463613    0.017842\n",
       "std    1.466917e+07     2529.492654       0.041700    0.013975\n",
       "min    1.922512e+07    24322.000000       0.098566    0.004829\n",
       "25%    4.734575e+07    31190.000000       0.442598    0.009936\n",
       "50%    5.771337e+07    32549.000000       0.463419    0.013565\n",
       "75%    6.817859e+07    33812.000000       0.485823    0.019978\n",
       "max    1.140157e+08    54681.000000       0.599044    0.105110"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "qc_hist(qc_tr_out, qc_te_out, \"libsize\", thr=(qc_thresholds[\"libsize_lo\"], qc_thresholds[\"libsize_hi\"]))\n",
    "qc_hist(qc_tr_out, qc_te_out, \"detected_genes\", thr=(qc_thresholds[\"detected_lo\"], qc_thresholds[\"detected_hi\"]))\n",
    "qc_hist(qc_tr_out, qc_te_out, \"zero_fraction\", thr=(qc_thresholds[\"zero_lo\"], qc_thresholds[\"zero_hi\"]))\n",
    "qc_hist(qc_tr_out, qc_te_out, \"top1_frac\", thr=(qc_thresholds[\"top1_lo\"], qc_thresholds[\"top1_hi\"]))\n",
    "\n",
    "display(qc_tr_out[[\"libsize\",\"detected_genes\",\"zero_fraction\",\"top1_frac\",\"keep\"]].describe())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ce9e2c6",
   "metadata": {},
   "source": [
    "## QC 4 SCATTER POUR LES OUTLIERS\n",
    "libsize vs detected_genes, les couleurs : conversé vs filtré\n",
    "\n",
    "Pour visualisé les outliers rapidement:\n",
    "pour chaque échantillons on a en x-axis, le nombre total de read RNA-seq, le nombre de gènes avec counts > 0\n",
    "-> chaque point est un échantillon.\n",
    "La relation attendue est plus au moins linéaire et croissante <=> plus on séquence plus on a de counts."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "id": "afbb21cd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'./tcga_brca_qc_report/qc_scatter_libsize_vs_detected_genes_test.png'"
      ]
     },
     "execution_count": 165,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "qc_scatter(qc_tr_out, \"libsize\", \"detected_genes\", title=\"train\")\n",
    "qc_scatter(qc_te_out, \"libsize\", \"detected_genes\", title=\"test\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c6b3c92",
   "metadata": {},
   "source": [
    "# FILTRAGE GENE ET NORMALISATION \n",
    "## FILTRER LES ECHANTILLONS APRES QC\n",
    "On supprime les samples rejetés par QC, on met à jour counts et meta\n",
    "\n",
    "On peut donc travaillé sur des données propres"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "id": "2dcdc7ff",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(60660, 814) (60660, 208)\n"
     ]
    }
   ],
   "source": [
    "train_kept = qc_tr_out.index[qc_tr_out[\"keep\"]].tolist()\n",
    "test_kept  = qc_te_out.index[qc_te_out[\"keep\"]].tolist()\n",
    "\n",
    "counts_tr_f = counts_tr[train_kept]\n",
    "counts_te_f = counts_te[test_kept]\n",
    "\n",
    "meta_tr_f = meta_tr.set_index(SAMPLE_COL).loc[train_kept].reset_index()\n",
    "meta_te_f = meta_te.set_index(SAMPLE_COL).loc[test_kept].reset_index()\n",
    "\n",
    "print(counts_tr_f.shape, counts_te_f.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c3517ef",
   "metadata": {},
   "source": [
    "## FILTRER GENES SUR TRAIN \n",
    "On garde les gènes exprimés dans au moins ≥ X% des samples train\n",
    "\n",
    "Pour supprimer le bruit et réduire la dimension après"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "id": "8f56b1ee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genes kept: 43638 sur 60660 totaux\n"
     ]
    }
   ],
   "source": [
    "min_frac = 0.10\n",
    "min_n = max(5, int(np.ceil(min_frac * counts_tr_f.shape[1])))\n",
    "\n",
    "gene_keep = (counts_tr_f > 0).sum(axis=1) >= min_n\n",
    "\n",
    "counts_tr_fg = counts_tr_f.loc[gene_keep].copy()\n",
    "counts_te_fg = counts_te_f.loc[gene_keep].copy()\n",
    "\n",
    "print(\"Genes kept:\", int(gene_keep.sum()), \"sur\", counts_tr_f.shape[0], \"totaux\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e40b96c",
   "metadata": {},
   "source": [
    "# PCA sur train uniquemt\n",
    "## sélection des top variant genes sur train\n",
    "On fait la variance calculé sur le train et on sélectionne les top gènes\n",
    "\n",
    "Pour avoir une réduction dimensionnelle biologiqmt informée\n",
    "\n",
    "On fait une sélection des top 10k genes, et on ajoute un bloc pour l'exportation et analyse de ces gènes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "id": "8e2fd54d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(814, 10000) (208, 10000)\n"
     ]
    }
   ],
   "source": [
    "top_genes = min(10000, counts_tr_fg.shape[0])\n",
    "gene_var = counts_tr_fg.var(axis=1).sort_values(ascending=False)\n",
    "sel_genes = gene_var.head(top_genes).index\n",
    "\n",
    "Xtr_sel = counts_tr_fg.loc[sel_genes].T  # samples x genes\n",
    "Xte_sel = counts_te_fg.loc[sel_genes].T\n",
    "\n",
    "print(Xtr_sel.shape, Xte_sel.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "723e35a3",
   "metadata": {},
   "source": [
    "# Exporter des HVG (10k) protein_coding\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "id": "d331ea94",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved: hvg_top_300_protein_coding.csv hvg_top_300_protein_coding_gene_list.txt\n"
     ]
    }
   ],
   "source": [
    "df_pc = gene_var.reset_index()\n",
    "df_pc.columns = [\"ensembl_id\", \"variance\"]\n",
    "\n",
    "df_pc = df_pc.merge(\n",
    "    genes[[\"gene_name\", \"gene_type\"]].reset_index().rename(columns={\"index\": \"ensembl_id\"}),\n",
    "    on=\"ensembl_id\",\n",
    "    how=\"left\"\n",
    ")\n",
    "\n",
    "TOP_N_PC = 300   # pour l'exportation, on cherche: Quels sont les gènes protein_coding les plus variables du dataset ?\n",
    "df_pc_top = df_pc.head(TOP_N_PC)\n",
    "\n",
    "df_pc_top.to_csv(\n",
    "    os.path.join(OUTDIR, f\"hvg_top_{TOP_N_PC}_protein_coding.csv\"),\n",
    "    index=False\n",
    ")\n",
    "\n",
    "gene_list_pc = df_pc_top[\"gene_name\"].fillna(df_pc_top[\"ensembl_id\"]).astype(str).tolist()\n",
    "with open(os.path.join(OUTDIR, f\"hvg_top_{TOP_N_PC}_protein_coding_gene_list.txt\"), \"w\") as f:\n",
    "    f.write(\"\\n\".join(gene_list_pc))\n",
    "\n",
    "print(\"Saved:\",\n",
    "      f\"hvg_top_{TOP_N_PC}_protein_coding.csv\",\n",
    "      f\"hvg_top_{TOP_N_PC}_protein_coding_gene_list.txt\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b4473e6b",
   "metadata": {},
   "source": [
    "## Standardisation et PCA en fit sur train et transform sur Test\n",
    "On fait un StandardScaler et PCA sur le train."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "id": "248e2f12",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Explained variance PC1-5: [0.18980305 0.06868443 0.05816674 0.03684524 0.02278491]  | sum PC1-20: 0.5124001669094745\n",
      "Ztr: (814, 20)\n",
      "Zte: (208, 20)\n"
     ]
    }
   ],
   "source": [
    "scaler = StandardScaler()\n",
    "Xtr_z = scaler.fit_transform(Xtr_sel)\n",
    "Xte_z = scaler.transform(Xte_sel)\n",
    "\n",
    "pca = PCA(n_components=20, random_state=RANDOM_STATE)\n",
    "Ztr = pca.fit_transform(Xtr_z)\n",
    "Zte = pca.transform(Xte_z)\n",
    "\n",
    "expl = pca.explained_variance_ratio_\n",
    "print(\"Explained variance PC1-5:\", expl[:5], \" | sum PC1-20:\", expl[:20].sum())\n",
    "print(\"Ztr:\", Ztr.shape)\n",
    "print(\"Zte:\", Zte.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b5f0899",
   "metadata": {},
   "source": [
    "## Les plots\n",
    "On fait une PCA avec PC1 vs PC2 et train vs test\n",
    "\n",
    "Pour vérifier la structure globale, le batch effect et les outliers persistant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "id": "ff72c412",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'./tcga_brca_qc_report/pca_train_vs_test.png'"
      ]
     },
     "execution_count": 171,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plt.figure()\n",
    "plt.scatter(Ztr[:,0], Ztr[:,1], s=14, alpha=0.7, label=\"train\")\n",
    "plt.scatter(Zte[:,0], Zte[:,1], s=14, alpha=0.7, label=\"test\")\n",
    "plt.xlabel(\"PC1\"); plt.ylabel(\"PC2\")\n",
    "plt.legend(frameon=False)\n",
    "savefig(\"pca_train_vs_test.png\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09da810f",
   "metadata": {},
   "source": [
    "# METADATA\n",
    "## SELECTION AUTO DES COLONNES CATEGORIELLES ET NUMERIQUES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "id": "c1b89872",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Categorical candidates: ['specimen_type', 'preservation_method', 'synchronous_malignancy', 'ajcc_pathologic_stage', 'laterality', 'tissue_or_organ_of_origin', 'prior_malignancy', 'prior_treatment', 'diagnosis_is_primary_disease', 'method_of_diagnosis', 'ajcc_staging_system_edition', 'ajcc_pathologic_t', 'ajcc_pathologic_n', 'ajcc_pathologic_m', 'classification_of_tumor', 'icd_10_code', 'site_of_resection_or_biopsy', 'metastasis_at_diagnosis', 'follow_ups_disease_response', 'race', 'gender', 'ethnicity', 'vital_status', 'age_is_obfuscated', 'country_of_residence_at_enrollment', 'paper_Included_in_previous_marker_papers', 'paper_vital_status', 'paper_pathologic_stage', 'paper_BRCA_Pathology', 'paper_BRCA_Subtype_PAM50', 'paper_CNV.Clusters', 'paper_Mutation.Clusters', 'paper_DNA.Methylation.Clusters', 'paper_mRNA.Clusters', 'paper_miRNA.Clusters', 'paper_lncRNA.Clusters', 'paper_Protein.Clusters', 'paper_PARADIGM.Clusters', 'paper_Pan.Gyn.Clusters']\n",
      "Numeric candidates: ['days_to_collection', 'initial_weight', 'days_to_diagnosis', 'age_at_diagnosis', 'year_of_diagnosis', 'days_to_last_follow_up', 'age_at_index', 'days_to_birth', 'days_to_death', 'releasable', 'released', 'days_to_sample_procurement', 'year_of_birth', 'figo_staging_edition_year', 'paper_days_to_birth', 'paper_days_to_death', 'paper_days_to_last_followup', 'paper_age_at_initial_pathologic_diagnosis', 'paper_Tumor_Grade', 'paper_MSI_status', 'paper_HPV_Status', 'paper_tobacco_smoking_history']\n"
     ]
    }
   ],
   "source": [
    "def pick_categorical_cols(df, max_unique=30):\n",
    "    cols = []\n",
    "    for c in df.columns:\n",
    "        if c in [SAMPLE_COL, PATIENT_COL]:\n",
    "            continue\n",
    "        nun = df[c].nunique(dropna=True)\n",
    "        if nun >= 2 and nun <= max_unique and df[c].dtype == \"object\":\n",
    "            cols.append(c)\n",
    "    return cols\n",
    "\n",
    "def pick_numeric_cols(df):\n",
    "    cols = []\n",
    "    for c in df.columns:\n",
    "        if c in [SAMPLE_COL, PATIENT_COL]:\n",
    "            continue\n",
    "        if pd.api.types.is_numeric_dtype(df[c]):\n",
    "            cols.append(c)\n",
    "    return cols\n",
    "\n",
    "cat_cols = pick_categorical_cols(meta_tr_f, max_unique=25)\n",
    "num_cols = pick_numeric_cols(meta_tr_f)\n",
    "\n",
    "print(\"Categorical candidates:\", cat_cols)\n",
    "print(\"Numeric candidates:\", num_cols)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e33aad38",
   "metadata": {},
   "source": [
    "## BARPLOT (CATEGO) ET HISTO (NUMERIC)\n",
    "### uncomment plt show to see the figures directly (carefull, there are many of them)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "id": "d98cf791",
   "metadata": {},
   "outputs": [],
   "source": [
    "def barplot_top(df_tr, df_te, col, k=15):\n",
    "    tr = df_tr[col].astype(str).fillna(\"NA\").value_counts().head(k)\n",
    "    te = df_te[col].astype(str).fillna(\"NA\").value_counts().head(k)\n",
    "\n",
    "    cats = sorted(set(tr.index) | set(te.index))\n",
    "    trv = [tr.get(c, 0) for c in cats]\n",
    "    tev = [te.get(c, 0) for c in cats]\n",
    "\n",
    "    x = np.arange(len(cats))\n",
    "    plt.figure(figsize=(max(6, len(cats)*0.5), 4))\n",
    "    plt.bar(x-0.2, trv, width=0.4, label=\"train\")\n",
    "    plt.bar(x+0.2, tev, width=0.4, label=\"test\")\n",
    "    plt.xticks(x, cats, rotation=45, ha=\"right\")\n",
    "    plt.ylabel(\"n\"); plt.title(col)\n",
    "    plt.legend(frameon=False)\n",
    "    #plt.show()\n",
    "    return savefig(f\"meta_bar_{col}.png\")\n",
    "\n",
    "def hist_numeric(train_df, test_df, col, bins=60):\n",
    "    plt.figure()\n",
    "    plt.hist(pd.to_numeric(train_df[col], errors=\"coerce\").dropna(), bins=bins, alpha=0.6, label=\"train\")\n",
    "    plt.hist(pd.to_numeric(test_df[col], errors=\"coerce\").dropna(), bins=bins, alpha=0.6, label=\"test\")\n",
    "    plt.xlabel(col); plt.ylabel(\"n\")\n",
    "    plt.legend(frameon=False)\n",
    "    #plt.show()\n",
    "    return savefig(f\"meta_hist_{col}.png\")\n",
    "\n",
    "# Plots (limite volontaire pour éviter 200 figures)\n",
    "for c in cat_cols[:8]:\n",
    "    barplot_top(meta_tr_f, meta_te_f, c)\n",
    "\n",
    "for c in num_cols[:8]:\n",
    "    hist_numeric(meta_tr_f, meta_te_f, c)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24b95b8e",
   "metadata": {},
   "source": [
    "# SAUVEGARDE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "id": "d2730ab9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✓ Saved 10000 gene names to gene_names.npy\n",
      "  First 10 genes: ['COL1A1' 'FN1' 'IGHG1' 'COL1A2' 'MT-CO1' 'CPB1' 'COL3A1' 'MT-RNR2'\n",
      " 'MT-ND4' 'IGKC']\n",
      "  Example genes: ['COL1A1' 'LUM' 'SLC7A5' 'RAN' 'TGFB1']\n",
      "Saved to: ./tcga_brca_qc_report\n"
     ]
    }
   ],
   "source": [
    "qc_tr_out.to_csv(os.path.join(OUTDIR, \"qc_train.csv\"))\n",
    "qc_te_out.to_csv(os.path.join(OUTDIR, \"qc_test.csv\"))\n",
    "pd.Series(qc_thresholds).to_csv(os.path.join(OUTDIR, \"qc_thresholds_train_derived.csv\"))\n",
    "\n",
    "meta_tr_f.to_csv(os.path.join(OUTDIR, \"meta_train_postQC.csv\"), index=False)\n",
    "meta_te_f.to_csv(os.path.join(OUTDIR, \"meta_test_postQC.csv\"), index=False)\n",
    "\n",
    "# Matrices prêtes ML \n",
    "Xtr_sel.to_csv(os.path.join(OUTDIR, \"X_train_logCPM_selectedGenes.csv\"))\n",
    "Xte_sel.to_csv(os.path.join(OUTDIR, \"X_test_logCPM_selectedGenes.csv\"))\n",
    "\n",
    "np.save(os.path.join(OUTDIR, \"X_train_z.npy\"), Xtr_z)\n",
    "np.save(os.path.join(OUTDIR, \"X_test_z.npy\"), Xte_z)\n",
    "\n",
    "\n",
    "# Extraire les vrais noms de gènes depuis sel_genes (ensembl IDs) et les mapper aux gene_name dans brca_genes.csv\n",
    "# sel_genes contient les ensembl IDs des 10,000 gènes sélectionnés\n",
    "selected_gene_names = []\n",
    "\n",
    "for ensembl_id in sel_genes:\n",
    "    # Chercher le gene_name correspondant dans genes DataFrame\n",
    "    if ensembl_id in genes.index:\n",
    "        gene_name = genes.loc[ensembl_id, 'gene_name']\n",
    "        # Si gene_name est NaN, utiliser l'ensembl_id\n",
    "        if pd.isna(gene_name):\n",
    "            selected_gene_names.append(ensembl_id)\n",
    "        else:\n",
    "            selected_gene_names.append(str(gene_name))\n",
    "    else:\n",
    "        # Si pas trouvé, utiliser l'ensembl_id\n",
    "        selected_gene_names.append(ensembl_id)\n",
    "\n",
    "# Convertir en array numpy\n",
    "gene_names_array = np.array(selected_gene_names, dtype=object)\n",
    "\n",
    "# Sauvegarder\n",
    "np.save(os.path.join(OUTDIR, \"gene_names.npy\"), gene_names_array)\n",
    "\n",
    "print(f\"✓ Saved {len(gene_names_array)} gene names to gene_names.npy\")\n",
    "print(f\"  First 10 genes: {gene_names_array[:10]}\")\n",
    "print(f\"  Example genes: {gene_names_array[[0, 100, 500, 1000, 5000]]}\")\n",
    "\n",
    "print(\"Saved to:\", OUTDIR)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f452a79",
   "metadata": {},
   "source": [
    "# GO enrichment \n",
    "On le fais sur gene_type dans brca_genes.csv, car GO beaucoup interprétable sur des gènes codants, et es gènes très variables peuvent inclure des lncRNA/pseudogenes qui ajoutent du bruit pour GO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "id": "5c20b00e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved: ./tcga_brca_qc_report/go_hvg_enrichment.csv\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    import gseapy as gp\n",
    "\n",
    "    GO_LIBRARY = \"GO_Biological_Process_2021\"\n",
    "    go_outdir = os.path.join(OUTDIR, \"go_hvg_enrichr\")\n",
    "    os.makedirs(go_outdir, exist_ok=True)\n",
    "\n",
    "    enr = gp.enrichr(\n",
    "        gene_list=gene_list_pc,     \n",
    "        gene_sets=[GO_LIBRARY],\n",
    "        organism=\"Human\",\n",
    "        outdir=go_outdir,\n",
    "        cutoff=0.05\n",
    "    )\n",
    "\n",
    "    if hasattr(enr, \"results\") and enr.results is not None:\n",
    "        enr.results.to_csv(os.path.join(OUTDIR, \"go_hvg_enrichment.csv\"), index=False)\n",
    "        print(\"Saved:\", os.path.join(OUTDIR, \"go_hvg_enrichment.csv\"))\n",
    "    else:\n",
    "        print(\"GO enrichment returned no results.\")\n",
    "\n",
    "except Exception as e:\n",
    "    print(\"GO enrichment skipped/failed:\", repr(e))\n",
    "    print(\"If no internet or gseapy not installed, this is expected.\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "tf-env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
